engine/src/main/java/org/camunda/bpm/engine/impl/batch/variables/BatchSetVariablesHandler.java

```diff
@@ -0,0 +1,116 @@
+ */
+package org.camunda.bpm.engine.impl.batch.variables;
+
+import org.camunda.bpm.engine.batch.Batch;
+import org.camunda.bpm.engine.impl.batch.AbstractBatchJobHandler;
+import org.camunda.bpm.engine.impl.batch.BatchConfiguration;
+import org.camunda.bpm.engine.impl.batch.BatchJobConfiguration;
+import org.camunda.bpm.engine.impl.batch.BatchJobContext;
+import org.camunda.bpm.engine.impl.batch.BatchJobDeclaration;
+import org.camunda.bpm.engine.impl.interceptor.CommandContext;
+import org.camunda.bpm.engine.impl.jobexecutor.JobDeclaration;
+import org.camunda.bpm.engine.impl.json.JsonObjectConverter;
+import org.camunda.bpm.engine.impl.persistence.entity.ByteArrayEntity;
+import org.camunda.bpm.engine.impl.persistence.entity.ExecutionEntity;
+import org.camunda.bpm.engine.impl.persistence.entity.MessageEntity;
+import org.camunda.bpm.engine.impl.persistence.entity.VariableInstanceEntity;
+import org.camunda.bpm.engine.variable.value.TypedValue;
+
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+
+public class BatchSetVariablesHandler extends AbstractBatchJobHandler<BatchConfiguration> {
+
+  public static final BatchJobDeclaration JOB_DECLARATION =
+      new BatchJobDeclaration(Batch.TYPE_SET_VARIABLES);
+
+  @Override
+  public void execute(BatchJobConfiguration configuration,
+                      ExecutionEntity execution,
+                      CommandContext commandContext,
+                      String tenantId) {
+
+    String byteArrayId = configuration.getConfigurationByteArrayId();
+    ByteArrayEntity byteArray = findByteArrayById(byteArrayId, commandContext);
+
+    byte[] configurationByteArray = byteArray.getBytes();
+    BatchConfiguration batchConfiguration = readConfiguration(configurationByteArray);
+
+    String batchId = batchConfiguration.getBatchId();
+    List<VariableInstanceEntity> variableInstances = commandContext.getVariableInstanceManager()
+        .selectByBatchId(batchId);
+
+    Map<String, Object> variables = new HashMap<>();
+    variableInstances.forEach(variableInstance -> {
+      String variableName = variableInstance.getName();
+      TypedValue variableValue = variableInstance.getTypedValue();
+
+      variables.put(variableName, variableValue);
+    });
+
+    List<String> processInstanceIds = batchConfiguration.getIds();
+    for (String processInstanceId : processInstanceIds) {
+      boolean initialLegacyRestrictions =
+          commandContext.isRestrictUserOperationLogToAuthenticatedUsers();
+
+      commandContext.disableUserOperationLog();
+      commandContext.setRestrictUserOperationLogToAuthenticatedUsers(true);
+
+      try {
+        commandContext.getProcessEngineConfiguration()
+            .getRuntimeService()
+            .setVariables(processInstanceId, variables);
+
+      } finally {
+        commandContext.enableUserOperationLog();
+        commandContext.setRestrictUserOperationLogToAuthenticatedUsers(initialLegacyRestrictions);
```

Comment:
Could we handle the context outside of the loop, so we don't do this for every instance?

Reply:
👍